app/js/email_confirmation_controller.js

```diff
@@ -1,15 +1,63 @@
 'use strict';
 
-/* global Utils */
+/* global User */
 
 (function(exports) {
+  var confirmation_poll_timeout;
+  var tmp_user;
+
+  function check_for_email_confirmation() {
+    if (tmp_user) {
+      return User.is_active(tmp_user.username).then(function (active) {
+        if (active) {
+          User.authenticate_temporary_user();
+          window.parent.Navigation.close_current_view();
+        }
+        return active;
+      });
+    }
+    return true;
+  }
+
+  function poll_for_email_confirmation() {
+    confirmation_poll_timeout = window.setTimeout(function() {
+      var promise = check_for_email_confirmation();
+
+      promise.then(function(active) {
+        if (!active) {
+          poll_for_email_confirmation();
+        }
+      });
+    }, 5000);
+  }
+
+  function handle_visibility_change() {
+    if (!document.hidden) {
+      if (confirmation_poll_timeout) {
+        clearTimeout(confirmation_poll_timeout);
+      }
+      poll_for_email_confirmation();
+    }
+  }
+
   var EmailConfirmationController = {
     init: function() {
-      var params = Utils.get_url_parameters(location);
-      if (params.email) {
+      var cancel_button = document.getElementById('cancel_button');
+      cancel_button.addEventListener('click', function() {
+        User.clear_temporary_user();
+        window.parent.Navigation.close_current_view();
```

Comment:
We should probably wait for the temporary user to be cleaned before closing the view.
```js
User.clear_temporary_user().then(function() {
  window.parent.Navigation.close_current_view();
});
```
I don't think that creates bugs but that sounds like good prevention against races.